src/ucp/core/ucp_listener.c

```diff
@@ -90,83 +104,99 @@ static int ucp_listener_remove_filter(const ucs_callbackq_elem_t *elem,
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    accept->addr->listener    = listener;
+    accept->addr->conn_req_id = id;
+    memcpy(&accept->addr->priv_addr, client_data, length);
+
+    return UCS_OK;
+}
+
+static ucs_status_t
+ucp_listener_conn_request_callback(void *arg, void *id,
+                                   const void *conn_priv_data, size_t length)
 {
     const ucp_wireup_sockaddr_priv_t *client_data = conn_priv_data;
     ucp_listener_h listener                       = arg;
-    ucp_unpacked_address_t remote_address;
     ucp_listener_accept_t *accept;
+    ucp_worker_h worker;
     uct_worker_cb_id_t prog_id;
-    ucp_ep_params_t params;
     ucs_status_t status;
-    ucp_ep_h ep;
 
     ucs_trace("listener %p: got connection request", listener);
 
-    params.field_mask = UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE;
-    params.err_mode   = client_data->err_mode;
-
-    status = ucp_address_unpack(client_data + 1, &remote_address);
-    if (status != UCS_OK) {
-        goto err;
+    /* Defer wireup init and user's callback to be invoked from the main thread */
+    accept = ucs_malloc(sizeof(*accept), "ucp_listener_accept");
+    if (accept == NULL) {
+        ucs_error("failed to allocate listener accept context");
+        return UCS_ERR_NO_MEMORY;
     }
 
-    if (client_data->is_full_addr) {
-        /* create endpoint to the worker address we got in the private data */
-        status = ucp_ep_create_to_worker_addr(listener->wiface.worker, &params,
-                                              &remote_address, UCP_EP_CREATE_AM_LANE,
-                                              "listener", &ep);
-        if (status != UCS_OK) {
-            goto err_free_address;
-        }
+    worker = listener->wiface.worker;
+    UCP_THREAD_CS_ENTER_CONDITIONAL(&worker->mt_lock);
 
-    } else {
-        status = ucp_ep_create_sockaddr_aux(listener->wiface.worker,
-                                            &params, &remote_address,
-                                            &ep);
-        if (status != UCS_OK) {
-            goto err_free_address;
+    if (listener->cb) {
+        status = ucp_listener_ep_create(listener, client_data, accept);
+    } else if (listener->addr_cb) {
+        status = ucp_listener_ep_addr_create(listener, id, client_data, length,
+                                             accept);
+        if (status == UCS_OK) {
+            /* hold id */
+            status = UCS_INPROGRESS;
         }
-
-        /* the listener's ep should be aware of the sent address from the client */
-        ep->flags |= UCP_EP_FLAG_LISTENER;
+    } else {
+        ucs_assert((listener->cb == NULL) && (listener->addr_cb == NULL));
+        /* No callback, but create internal EP for backward compatibility */
+        status = ucp_listener_ep_create(listener, client_data, accept);
     }
 
-    ucp_ep_update_dest_ep_ptr(ep, client_data->ep_ptr);
+    UCP_THREAD_CS_EXIT_CONDITIONAL(&worker->mt_lock);
```

Comment:
dont need it in the end of the function?

Reply:
`uct_worker_progress_register_safe` and `ucp_worker_signal_internal` are thread-safe